### PR TITLE
143370 - Change related to additional param (From Boolean to String)

### PR DESCRIPTION
**Pull request by ratnas104142**
This PR consists changes of [PBI 143370](https://dev.azure.com/ukhydro/Exchange%20Set%20Service/_workitems/edit/143370) as follows:

- Code changes to change the additional param from Boolean to String 'exchangeSetStandard'
- Added new test cases and updated existing unit tests
- Updated and added new functional tests

### CODE REVIEW

**Review by shirinbt14926 — COMMENTED** [review]
Few minor comments

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ExchangeSetGenerateFilesForProductIdentifierWithParameterexchangeSetStandardAsS57.cs [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+using System.IO;
+using System.Threading.Tasks;
+using UKHO.ExchangeSetService.API.FunctionalTests.Helper;
+using UKHO.ExchangeSetService.API.FunctionalTests.Models;
+using System.Collections.Generic;
+
+namespace UKHO.ExchangeSetService.API.FunctionalTests.FunctionalTests
+{
+    [TestFixture]
+    public class ExchangeSetGenerateFilesForProductIdentifierWithParameterexchangeSetStandardAsS57 : ObjectStorage
```
[the remark this answers is not in the document]
Comment (Arit14975) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ExchangeSetGenerateFilesForProductIdentifierWithParameterexchangeSetStandardAsS57.cs:12]: Hi Shirin, This is a good spot. We are trying to be consistent, as the parameter name itself starts with lower case, so we kept the same in FT.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ExchangeSetGenerateFilesForProductIdentifierWithParameterexchangeSetStandardAsS63.cs [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+using System.IO;
+using System.Threading.Tasks;
+using UKHO.ExchangeSetService.API.FunctionalTests.Helper;
+using UKHO.ExchangeSetService.API.FunctionalTests.Models;
+using System.Collections.Generic;
+
+namespace UKHO.ExchangeSetService.API.FunctionalTests.FunctionalTests
+{
+    [TestFixture]
+    public class ExchangeSetGenerateFilesForProductIdentifierWithParameterexchangeSetStandardAsS63 : ObjectStorage
```
[the remark this answers is not in the document]
Comment (Arit14975) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ExchangeSetGenerateFilesForProductIdentifierWithParameterexchangeSetStandardAsS63.cs:12]: Hi Shirin, This is a good spot. We are trying to be consistent, as the parameter name itself starts with lower case, so we kept the same in FT.

[the remark this answers is not in the document]
Comment (shirinbt14926) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ExchangeSetGenerateFilesForProductIdentifierWithParameterexchangeSetStandardAsS63.cs:12]: Thats also one good way to name it, no worries

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ExchangeSetGenerateFilesForProductIdentifierWithParameterexchangeSetStandardAsS57.cs [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+                //Verify README.TXT file content
+                FileContentHelper.CheckReadMeTxtFileContent(Path.Combine(downloadedFolderPath, Config.ExchangeSetEncRootFolder, Config.ExchangeReadMeFile));
+            }
+        }
+
+         //PBI 143370: Change related to additional param (From Boolean to String)
+        [Test]
+        [Category("QCOnlyTest-AIODisabled")]
+        public async Task WhenICallProductIdentifiersApiWithS57ForParameterexchangeSetStandard_ThenACatalogFileIsGenerated()
```
Comment (harshal11869) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ExchangeSetGenerateFilesForProductIdentifierWithParameterexchangeSetStandardAsS57.cs:70]: Can we change the method name as below & Class name should have s in it ProductIdentifier => ProductIdentifiers
```suggestion
        public async Task WhenICallProductIdentifiersApiWithS57ExchangeSetStandardParameter_ThenACatalogFileIsGenerated()
```
**_Note:_** Please check other classes and test methods same as this and do the required changes.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ExchangeSetGenerateFilesForProductIdentifierWithoutParameterexchangeSetStandard.cs [hunk 4]
```diff
@@ -9,25 +9,25 @@
 namespace UKHO.ExchangeSetService.API.FunctionalTests.FunctionalTests
 {
     [TestFixture]
-    public class ExchangeSetGenerateFilesForProductIdentifierWithIncorrectValueForIsUnencryptedParameter : ObjectStorage
+    public class ExchangeSetGenerateFilesForProductIdentifierWithoutParameterexchangeSetStandard : ObjectStorage
```
Comment (harshal11869) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ExchangeSetGenerateFilesForProductIdentifierWithoutParameterexchangeSetStandard.cs:12]: Please use Pascal Case e.g. exchangeSetStandard => ExchangeSetStandard wherever applicable. Please check the below change as below at other places as well.
```suggestion
    public class ExchangeSetGenerateFilesForProductIdentifiersWithoutExchangeSetStandardParameter : ObjectStorage
```

**Review by harshal11869 — CHANGES_REQUESTED** [review]

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ExchangeSetGenerateFilesForProductIdentifierWithParameterexchangeSetStandardAsS57.cs (hunk 3 above)
Comment (Arit14975) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ExchangeSetGenerateFilesForProductIdentifierWithParameterexchangeSetStandardAsS57.cs]: Resolved.Thanks

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ExchangeSetGenerateFilesForProductIdentifierWithoutParameterexchangeSetStandard.cs (hunk 4 above)
Comment (Arit14975) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ExchangeSetGenerateFilesForProductIdentifierWithoutParameterexchangeSetStandard.cs]: Resolved.Thanks

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/Helper/FileContentHelper.cs [hunk 5]
```diff
@@ -248,6 +245,10 @@ public static void CheckCatalogueFileNoContent(string inputFile, List<ProductVer
         public static void DeleteDirectory(string fileName)
         {
             string path = Path.GetTempPath();
+            if(Directory.Exists(Path.Combine(path, Config.BESSConfig.TempFolderName)))
+            {
+                Directory.Delete(Path.Combine(path, Config.BESSConfig.TempFolderName),true);
```
Comment (harshal11869) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/Helper/FileContentHelper.cs:250]:
```suggestion
                Directory.Delete(Path.Combine(path, Config.BESSConfig.TempFolderName), true);
```

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ValidateAuthorizationForAPIsWithExchangeSetStandardParameterTests.cs [hunk 6]
```diff
@@ -23,102 +22,79 @@ public void SetupAsync()
...
-            cleanUpBatchIdList.Add(fssBatchId);
         }
         #endregion
 
         #region ProductIdentifier Api
         // PBI 140109 : ESS API : Add authorization to allow only UKHO people to create unencrypted ES 
         [Test]
         [Category("SmokeTest-AIODisabled")]
-        public async Task WhenICallTheProductIdentifierApiWithAValidB2cTokenAndIsUnencryptedParameterAsTrue_ThenAnUnauthorizedResponseIsReturned()
+        public async Task WhenICallTheProductIdentifierApiWithAValidB2cTokenAnds57AsexchangeSetStandardParameter_ThenAnUnauthorizedResponseIsReturned()
```
Comment (harshal11869) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ValidateAuthorizationForAPIsWithExchangeSetStandardParameterTests.cs:48]: Please change Identifier to Identifiers wherever applicable
```suggestion
        public async Task WhenICallTheProductIdentifiersApiWithAValidB2cTokenAnds57AsexchangeSetStandardParameter_ThenAnUnauthorizedResponseIsReturned()
```

**Review by harshal11869 — CHANGES_REQUESTED** [review]

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/Helper/FileContentHelper.cs (hunk 5 above)
Comment (Arit14975) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/Helper/FileContentHelper.cs]: Resolved. Thanks

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ValidateAuthorizationForAPIsWithExchangeSetStandardParameterTests.cs (hunk 6 above)
Comment (Arit14975) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ValidateAuthorizationForAPIsWithExchangeSetStandardParameterTests.cs]: Resolved. Thanks

**Review by Nadeem14456 — CHANGES_REQUESTED** [review]

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ValidateAuthorizationForAPIsWithExchangeSetStandardParameterTests.cs [hunk 7]
```diff
@@ -23,102 +22,79 @@ public void SetupAsync()
         // PBI 140109 : ESS API : Add authorization to allow only UKHO people to create unencrypted ES 
         [Test]
         [Category("SmokeTest-AIODisabled")]
-        public async Task WhenICallTheSinceDateTimeApiWithAValidB2cTokenAndIsUnencryptedParameterAsTrue_ThenAnUnauthorizedResponseIsReturned()
+        public async Task WhenICallSinceDateTimeApiWithAValidB2cTokenAnds57AsexchangeSetStandardParameter_ThenAnUnauthorizedResponseIsReturned()
```
Comment (harshal11869) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ValidateAuthorizationForAPIsWithExchangeSetStandardParameterTests.cs:25]: Please change wherever applicable in FTs
```suggestion
        public async Task WhenICallSinceDateTimeApiWithAValidB2cTokenAndS57AsExchangeSetStandardParameter_ThenAnUnauthorizedResponseIsReturned()
```

**Review by harshal11869 — CHANGES_REQUESTED** [review]

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API/Filters/BespokeExchangeSetAuthorizationFilterAttribute.cs [hunk 8]
```diff
@@ -14,7 +15,7 @@ namespace UKHO.ExchangeSetService.API.Filters
     public class BespokeExchangeSetAuthorizationFilterAttribute : ActionFilterAttribute
     {
         private const string TokenAudience = "aud";
-        private const string IsUnencrypted = "IsUnencrypted";
+        private const string ExchangeSetStandard = "exchangeSetStandard";
         private readonly IOptions<AzureADConfiguration> azureAdConfiguration;
 
         public BespokeExchangeSetAuthorizationFilterAttribute(IOptions<AzureADConfiguration> azureAdConfiguration)
```
[the remark this answers is not in the document]
Comment (ratnas104142) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API/Filters/BespokeExchangeSetAuthorizationFilterAttribute.cs:23]: Resolved!

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ValidateAuthorizationForAPIsWithExchangeSetStandardParameterTests.cs (hunk 7 above)
Comment (Arit14975) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ValidateAuthorizationForAPIsWithExchangeSetStandardParameterTests.cs]: Resolved. Thanks

**Review by shirinbt14926 — COMMENTED** [review]
Minor comments please

**Review by harshal11869 — APPROVED** [review]
Looks Good. Only one comment need to be resolved. Thanks !!!

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ExchangeSetGenerateFilesForProductVersionWithS63ExchangeSetStandardParameter.cs [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+                cleanUpBatchIdList.Add(batchId);
+                var downloadFolderPath = await FileContentHelper.CreateExchangeSetFile(apiResponse, FssJwtToken);
+                downloadedFolderPathList.Add(downloadFolderPath);
+            }
+        }
+
+         //PBI 143370: Change related to additional param (From Boolean to String)
+        [Test]
+        [Category("QCOnlyTest-AIODisabled")]
+        public async Task WhenICallProductVersionApiWithS63ExchangeSetStandardParameter_ThenAProductTxtFileIsGenerated()
```
[the remark this answers is not in the document]
Comment (Arit14975) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ExchangeSetGenerateFilesForProductVersionWithS63ExchangeSetStandardParameter.cs:39]: Done. Thanks

**Review by shirinbt14926 — APPROVED** [review]
Thanks @ratnas104142  and @Arit14975 for making changes. 
LGTM.

**Review by Mayuresh-Gawde — APPROVED** [review]
LGTM from QA!!

**Review by Nadeem14456 — APPROVED** [review]
LGTM.

**Review by DanOwenUKHO — CHANGES_REQUESTED** [review]
One small change

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API/Filters/BespokeExchangeSetAuthorizationFilterAttribute.cs [hunk 10]
```diff
@@ -14,31 +16,62 @@ namespace UKHO.ExchangeSetService.API.Filters
...
                 }
             }
             await next();
         }
+
+        private static bool ValidateExchangeSetStandard<TEnum>(string exchangeSetStandard, out TEnum result) where TEnum : struct, Enum
+        {
+            result = default;
+            if (string.IsNullOrEmpty(exchangeSetStandard))
+                return false;
```
[the remark this answers is not in the document]
Comment (shirinbt14926) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API/Filters/BespokeExchangeSetAuthorizationFilterAttribute.cs:62]: Done, thanks

**Pull request merged**